src/Datadog.Trace.ClrProfiler.Managed/Integrations/Testing/XUnitIntegration.cs

```diff
@@ -0,0 +1,534 @@
+            long moduleVersionPtr)
+        {
+            if (xunitTestAssemblyRunner == null) { throw new ArgumentNullException(nameof(xunitTestAssemblyRunner)); }
+
+            Type xunitTestAssemblyRunnerType = xunitTestAssemblyRunner.GetType();
+            Func<object, object, object, object, object, object> executeAsync;
+
+            try
+            {
+                executeAsync =
+                    MethodBuilder<Func<object, object, object, object, object, object>>
+                       .Start(moduleVersionPtr, mdToken, opCode, XUnitRunTestCollectionAsyncMethod)
+                       .WithConcreteType(xunitTestAssemblyRunnerType)
+                       .WithParameters(messageBus, testCollection, testCases, cancellationTokenSource)
+                       .WithNamespaceAndNameFilters(ClrNames.GenericTask, "Xunit.Sdk.IMessageBus", "Xunit.Abstractions.ITestCollection", "System.Collections.Generic.IEnumerable`1<T>", "System.Threading.CancellationTokenSource")
+                       .Build();
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorRetrievingMethod(
+                    exception: ex,
+                    moduleVersionPointer: moduleVersionPtr,
+                    mdToken: mdToken,
+                    opCode: opCode,
+                    instrumentedType: XUnitTestAssemblyRunnerType,
+                    methodName: XUnitRunTestCollectionAsyncMethod,
+                    instanceType: xunitTestAssemblyRunnerType.AssemblyQualifiedName);
+                throw;
+            }
+
+            object returnValue = null;
+            Exception exception = null;
+            try
+            {
+                returnValue = executeAsync(xunitTestAssemblyRunner, messageBus, testCollection, testCases, cancellationTokenSource);
+            }
+            catch (TargetInvocationException ex)
+            {
+                exception = ex.InnerException;
+                throw;
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
+                throw;
+            }
+            finally
+            {
+                returnValue = AsyncTool.AddContinuation<object>(
+                    returnValue,
+                    exception,
+                    null,
+                    async (r, ex, state) =>
+                    {
+                        // We have to ensure the flush of the buffer after we finish the tests of an assembly.
+                        // For some reason, sometimes when all test are finished none of the callbacks to handling the tracer disposal is triggered.
+                        // So the last spans in buffer aren't send to the agent.
+                        // Other times we reach the 500 items of the buffer in a sec and the tracer start to drop spans.
+                        // In a test scenario we must keep all spans.
+                        await Tracer.Instance.FlushAsync().ConfigureAwait(false);
```

Comment:
This concerns me a little because we would ideally like to have a strict boundary between integrations open/closing spans and the Tracer flushing spans.

@lucaspimentel / @colin-higgins  do you have any idea why this might be happening? Maybe we should be dedicating one thread to just flushing (instead of a series of async calls) and make it a non-Background thread so we make sure to flush once more before exiting.

Reply:
The problem here is no one of the Current domain events gets called when the test command finishes.

```csharp
// Register callbacks to make sure we flush the traces before exiting
AppDomain.CurrentDomain.ProcessExit += CurrentDomain_ProcessExit;
AppDomain.CurrentDomain.DomainUnload += CurrentDomain_DomainUnload;
AppDomain.CurrentDomain.UnhandledException += CurrentDomain_UnhandledException;
```

That's why I'm kind of forcing the flush when all tests of a given collection are executed. Without the flush, yeah I could do a Thead.Sleep / Task.Delay of 1 second so the tracer flushes automatically, but that's worst.

Reply:
I would be ok with doing this in test frameworks because they are a special case. They don't run code the way a normal app does so we can't rely on the events above. Also flushing at _any_ time should be safe. Only traces that are already finished and sent to the `AgentWriter` get flushed, not any in-flight traces. But if @zacharycmontoya has strong concerns we should consider our options.

Reply:
Makes sense @lucaspimentel , especially the flushing at any time part, thanks for following up. As long as this is only exposed internally, I'm on board. We can consider this concern resolved 👍